File: src/apps/daemon/internal/platform/mobilehost/mobile_tunnel_supervisor.go
package mobilehost

import (
	"errors"
	"sync"
)

type MobileTunnelState string

const (
	TunnelStopped      MobileTunnelState = "stopped"
	TunnelStarting     MobileTunnelState = "starting"
	TunnelRunning      MobileTunnelState = "running"
	TunnelPausing      MobileTunnelState = "pausing"
	TunnelReconnecting MobileTunnelState = "reconnecting"
	TunnelFailed       MobileTunnelState = "failed"
)

type MobileTunnelConfig struct {
	TunName          string
	Mtu              uint16
	IPv4Address      string
	IPv4Netmask      string
	DNSServers       []string
	IncludedPackages map[string]struct{}
	ExcludedPackages map[string]struct{}
}

type MobileTunnelSupervisor struct {
	mu               sync.RWMutex
	config           MobileTunnelConfig
	state            MobileTunnelState
	activeTunFd      int
	totalReconnects  uint32
}

func NewMobileTunnelSupervisor(config MobileTunnelConfig) *MobileTunnelSupervisor {
	if config.IncludedPackages == nil {
		config.IncludedPackages = make(map[string]struct{})
	}
	if config.ExcludedPackages == nil {
		config.ExcludedPackages = make(map[string]struct{})
	}
	return &MobileTunnelSupervisor{
		config:          config,
		state:           TunnelStopped,
		activeTunFd:     -1,
		totalReconnects: 0,
	}
}

func (s *MobileTunnelSupervisor) Start(tunFd int) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.state == TunnelRunning {
		return errors.New("tunnel already running")
	}
	s.activeTunFd = tunFd
	s.state = TunnelRunning
	return nil
}

func (s *MobileTunnelSupervisor) Stop() {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.state = TunnelStopped
	s.activeTunFd = -1
}

func (s *MobileTunnelSupervisor) TriggerReconnect() {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.totalReconnects++
	s.state = TunnelReconnecting
}

func (s *MobileTunnelSupervisor) IsPackageRouted(pkg string) bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	// 1. Excluded takes precedence
	if _, ok := s.config.ExcludedPackages[pkg]; ok {
		return false
	}
	// 2. If inclusion set is not empty, must be in inclusion set
	if len(s.config.IncludedPackages) > 0 {
		_, ok := s.config.IncludedPackages[pkg]
		return ok
	}
	// 3. Otherwise route all
	return true
}

func (s *MobileTunnelSupervisor) State() MobileTunnelState {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.state
}

func (s *MobileTunnelSupervisor) TunFd() int {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.activeTunFd
}

func (s *MobileTunnelSupervisor) ReconnectCount() uint32 {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.totalReconnects
}

File: src/packages/lumicore/src/platform/mobile_tunnel_supervisor.rs
//! # Universal Mobile Tunnel Supervisor
//!
//! Supervises mobile VPN TUN lifecycle, package-based split tunneling filtering,
//! MTU negotiation, and seamless reconnect states.

use std::collections::HashSet;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MobileTunnelState {
    Stopped,
    Starting,
    Running,
    Pausing,
    Reconnecting,
    Failed,
}

#[derive(Debug, Clone)]
pub struct MobileTunnelConfig {
    pub tun_name: String,
    pub mtu: u16,
    pub ipv4_address: String,
    pub ipv4_netmask: String,
    pub dns_servers: Vec<String>,
    pub included_packages: HashSet<String>,
    pub excluded_packages: HashSet<String>,
}

pub struct MobileTunnelSupervisor {
    config: MobileTunnelConfig,
    state: MobileTunnelState,
    active_tun_fd: Option<i32>,
    total_reconnects: u32,
}

impl MobileTunnelSupervisor {
    pub fn new(config: MobileTunnelConfig) -> Self {
        Self {
            config,
            state: MobileTunnelState::Stopped,
            active_tun_fd: None,
            total_reconnects: 0,
        }
    }

    pub fn start(&mut self, tun_fd: i32) -> Result<(), &'static str> {
        if self.state == MobileTunnelState::Running {
            return Err("Tunnel already running");
        }
        self.active_tun_fd = Some(tun_fd);
        self.state = MobileTunnelState::Running;
        Ok(())
    }

    pub fn stop(&mut self) {
        self.state = MobileTunnelState::Stopped;
        self.active_tun_fd = None;
    }

    pub fn trigger_reconnect(&mut self) {
        self.total_reconnects += 1;
        self.state = MobileTunnelState::Reconnecting;
    }

    pub fn is_package_routed(&self, package_name: &str) -> bool {
        // Excluded takes precedence
        if self.config.excluded_packages.contains(package_name) {
            return false;
        }
        // If inclusion list is not empty, must be in inclusion list
        if !self.config.included_packages.is_empty() {
            return self.config.included_packages.contains(package_name);
        }
        // By default, route all
        true
    }

    pub fn state(&self) -> MobileTunnelState {
        self.state
    }

    pub fn tun_fd(&self) -> Option<i32> {
        self.active_tun_fd
    }

    pub fn reconnect_count(&self) -> u32 {
        self.total_reconnects
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mobile_tunnel_lifecycle() {
        let mut inc = HashSet::new();
        inc.insert("com.android.chrome".to_string());
        let mut exc = HashSet::new();
        exc.insert("com.bank.app".to_string());

        let cfg = MobileTunnelConfig {
            tun_name: "tun0".to_string(),
            mtu: 1500,
            ipv4_address: "172.19.0.1".to_string(),
            ipv4_netmask: "255.255.255.0".to_string(),
            dns_servers: vec!["1.1.1.1".to_string()],
            included_packages: inc,
            excluded_packages: exc,
        };

        let mut sup = MobileTunnelSupervisor::new(cfg);
        assert_eq!(sup.state(), MobileTunnelState::Stopped);

        sup.start(10).unwrap();
        assert_eq!(sup.state(), MobileTunnelState::Running);
        assert_eq!(sup.tun_fd(), Some(10));

        assert!(sup.is_package_routed("com.android.chrome"));
        assert!(!sup.is_package_routed("com.bank.app"));
        assert!(!sup.is_package_routed("com.other.app")); // because included_packages is non-empty

        sup.trigger_reconnect();
        assert_eq!(sup.state(), MobileTunnelState::Reconnecting);
        assert_eq!(sup.reconnect_count(), 1);
    }
}
